### PR TITLE
ICB: Fix compilation errors for ptrdiff_t

### PR DESCRIPTION
**Pull request by n-i-x**
ICB: Fix compilation error for ptrdiff_t

```
../../../../engines/icb/common/ptr_util.cpp: In function ‘uint32 ICB::MemoryUtil::encodePtr(uint8*)’:
../../../../engines/icb/common/ptr_util.cpp:44:2: error: ‘ptrdiff_t’ was not declared in this scope
  ptrdiff_t diff = ptr - (uint8 *)0;
  ^
../../../../engines/icb/common/ptr_util.cpp:44:2: note: suggested alternative:
In file included from /usr/include/c++/5/new:39:0,
                 from ../../../../backends/platform/libretro/portdefs.h:32,
                 from ../../../../common/scummsys.h:44,
                 from ../../../../engines/icb/common/px_common.h:31,
                 from ../../../../engines/icb/common/ptr_util.cpp:28:
/usr/include/aarch64-linux-gnu/c++/5/bits/c++config.h:197:28: note:   ‘std::ptrdiff_t’
   typedef __PTRDIFF_TYPE__ ptrdiff_t;
                            ^
../../../../engines/icb/common/ptr_util.cpp:45:24: error: ‘diff’ was not declared in this scope
  ptrRef.ref = (uint32)(diff & 0xFFFFFFFF);
                        ^
Makefile:561: recipe for target 'engines/icb/common/ptr_util.o' failed
make: *** [engines/icb/common/ptr_util.o] Error 1
make: *** Waiting for unfinished jobs....
```

### CODE REVIEW

**Comment by aquadran** [comment]
ptrdiff_t is part of "stddef.h". looks like you have some issue with your compiler. also std:: is not allowed. rejecting.

**Comment by i30817** [comment]
This is happening again after i upgraded ubuntu. To be fair, that might have broken something, but isn't it necessary to `include <cstddef>' or something?

**Comment by i30817** [comment]
tl;dr: @aquadran , in order to compile in gcc 11 in ubuntu i had to `#include "stddef.h"` in that file. I suppose importing the scummsys.h file would work too, although that may be specific to the scum engine, not sure.

**Comment by lephilousophe** [comment]
Works fine here on Ubuntu 22.04.1 with GCC 11.
Inclusion chain is as following:
- `engines/icb/common/ptr_util.cpp` at line 27,
- `engines/icb/common/px_common.h` at line 30,
- `common/scummsys.h` at line 123,
- `/usr/lib/gcc/x86_64-linux-gnu/11/include/stddef.h`

Please fix your setup.

**Comment by lephilousophe** [comment]
AHA! Found it!
> `                 from ../../../../backends/platform/libretro/portdefs.h:32,`
You are using Retroarch. This is not ScummVM!
Report the bug there.
ScummVM team doesn't support Retroarch as it's full of patches not under our control.
As I now see it, the bug obviously comes from their patches.

**Comment by i30817** [comment]
I'm using a core that is not retroarch, and uses current upstream code : https://github.com/diablodiab/libretro-scummvm-backend

**Comment by n-i-x** [comment]
libretro is retroarch

**Comment by i30817** [comment]
Yes, but i mentioning specifically because there is almost no code drift. Apparently one of the imports got missing. Personally i find it a weird practice not to import the standard c library imports you need and get them 5 levels deep, so i reported it here in the hope that some reconsideration was done. It's just as easy to fix in the other place (especially because it's not dead like the actual retroarch scummvm core).

**Comment by lephilousophe** [comment]
Except for autogenerated code (Lex & Yacc) and some specific cases which may be removed, engines don't include system headers directly.
They are supposed to use the common/scummsys.h way to access C functions especially to be the more portable.
As for any fork it's the job of the one who forks to maintain its patches.

**Comment by sev-** [comment]
ScummVM is about code portability. We do not allow non-portable code outside of backends. That’s why. 

RetroArch has a track record of deliberately breaking that and ignoring our suggestions. 

Glad that this PR was promptly rejected

**Comment by i30817** [comment]
If you 'do not allow' that kind of code why is it that the _only_ error in _all engines_ occurs in this file when i compile with that backend?

**Comment by i30817** [comment]
Also:
```
i3@sleipnir:~/Documents/Projects/scummvm/engines$ grep -r  stddef.h . --exclude='*.o' --exclude='*.dwo' --exclude='*.a'
./hypno/grammar_arc.cpp:#   include <stddef.h> /* INFRINGES ON USER NAME SPACE */
./hypno/grammar_arc.cpp:#  include <stddef.h> /* INFRINGES ON USER NAME SPACE */
./hypno/grammar_mis.cpp:#   include <stddef.h> /* INFRINGES ON USER NAME SPACE */
./hypno/grammar_mis.cpp:#  include <stddef.h> /* INFRINGES ON USER NAME SPACE */
./sky/compact.cpp:#include <stddef.h>	// for ptrdiff_t
./private/grammar.cpp:#   include <stddef.h> /* INFRINGES ON USER NAME SPACE */
./private/grammar.cpp:#  include <stddef.h> /* INFRINGES ON USER NAME SPACE */
./icb/common/ptr_util.cpp:#include "stddef.h"
./director/lingo/lingo-gr.cpp:#   include <stddef.h> /* INFRINGES ON USER NAME SPACE */
./director/lingo/lingo-gr.cpp:#  include <stddef.h> /* INFRINGES ON USER NAME SPACE */

```
I guess they're all unstable engines anyway.

**Comment by lephilousophe** [comment]
As I already said, these files containing the include are autogenerated by bison.
So please stop arguing your change is needed.
Go fix the project you use instead of bothering the project which actually works.

**Comment by i30817** [comment]
I'm not arguing it's needed, i was just confused of why something was said that i saw was not true before it was said. If you're saying it's autogenerated, i believe it. No need to get like this.
